src/Equinox.MessageDb/MessageDb.fs

```diff
@@ -324,38 +188,42 @@ type MessageDbContext(client: MessageDbClient, batchOptions: BatchOptions) =
         let snapshotStream = Snapshot.streamName category streamId
-        let! _, events = Read.loadLastEvent log client.ReadRetryPolicy client.Reader requireLeader snapshotStream (Some eventType) ct
-        return Snapshot.decode tryDecode events }
+        let! _, events = Read.loadLastEvent client.ReadRetryPolicy client.Reader requireLeader snapshotStream (Some eventType) ct
+        let decoded = Snapshot.decode tryDecode events
+        let act = Activity.Current
+        if act <> null then
+          let version = match decoded with ValueNone -> -1L | ValueSome (t, _) -> t.version
+          act.SetTag("eqx.snapshot_version", version) |> ignore
+        return decoded }
 
-    member _.Reload(log, streamName, requireLeader, token, tryDecode, ct): Task<StreamToken * 'event[]> = task {
+    member _.Reload(streamName, requireLeader, token, tryDecode, fold, initial, ct): Task<struct(StreamToken * 'state)> = task {
         let streamVersion = Token.streamVersion token
         let startPos = streamVersion + 1L // Reading a stream uses {inclusive} positions, but the streamVersion is `-1`-based
-        let! version, events = Read.loadForwardsFrom log client.ReadRetryPolicy client.Reader batchOptions.BatchSize batchOptions.MaxBatches streamName startPos requireLeader ct
-        return Token.create (max streamVersion version), Array.chooseV tryDecode events }
+        let! version, state = Read.loadForwardsFrom fold initial tryDecode client.ReadRetryPolicy client.Reader batchOptions.BatchSize batchOptions.MaxBatches streamName startPos requireLeader ct
+        return struct(Token.create (max streamVersion version), state) }
 
-    member internal _.TrySync(log, category, streamId, streamName, token, encodedEvents: IEventData<EventBody>[], ct): Task<GatewaySyncResult> = task {
+    member internal _.TrySync(_category, _streamId, streamName, token, encodedEvents: IEventData<EventBody>[], ct): Task<GatewaySyncResult> = task {
         let streamVersion = Token.streamVersion token
-        match! Write.writeEvents log client.WriteRetryPolicy client.Writer (category, streamId, streamName) (StreamVersion streamVersion) encodedEvents ct with
+        match! Write.writeEvents client.WriteRetryPolicy client.Writer streamName (StreamVersion streamVersion) encodedEvents ct with
         | MdbSyncResult.Written version' ->
             let token = Token.create version'
             return GatewaySyncResult.Written token
         | MdbSyncResult.ConflictUnknown ->
             return GatewaySyncResult.ConflictUnknown }
 
-    member _.StoreSnapshot(log, category, streamId, event, ct) = task {
+    member _.StoreSnapshot(category, streamId, event, ct) = task {
         let snapshotStream = Snapshot.streamName category streamId
-        let category = Snapshot.snapshotCategory category
         let act = Activity.Current
         if act <> null then act.SetTag("eqx.snapshot_written", true) |> ignore
```

Comment:
is there an operation success indicator / is one required for the snapshot write completing (and might it prematurely set the other success indicator) - I guess a good reason to have a named extension method making it obvious

Reply:
> is one required for the snapshot write completing

I've been thinking about the semantics of snapshots. They're an optimisation, so if we fail to write a snapshot, should we just swallow it?

Reply:
Nah, nobody wants silent failures. If someone asked for it, maybe. But you'd probably want metrics to be able to extract the failure rate.
Also for people with streams that take seriously long to walk and/or fold, one might implement proactive snapshotting.
Having the snapshots there can be critical to meeting an SLA
And if it turns out that this is the very thing that's making your system unstable or actually ruining perf instead of improving it, you want to know

Reply:
> Also for people with streams that take seriously long to walk and/or fold, one might implement proactive snapshotting.

Yeah, I added a frequency parameter to the JS version which I think will see a lot of use, will contribute here when I get the time!

Reply:
also if a nullable varchar max was made available to abuse to store snapshots inside the main stream, you wouldnt suddenly want different semantics. Keep it direct and surprise free I say.